src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -145,6 +146,12 @@ public PerLeafResult searchLeaf(LeafReaderContext context, int k) throws IOExcep
             Map<Integer, Float> result = doExactSearch(context, new BitSetIterator(filterBitSet, cardinality), cardinality, k);
             return new PerLeafResult(filterWeight == null ? null : filterBitSet, result);
         }
+        /*
+         * If filters match all docs in this segment, then there is no need to do any extra step
+         * and should directly do ANN Search*/
+        if (filterWeight != null && cardinality == maxDoc) {
+            return new PerLeafResult(filterBitSet, doANNSearch(context, filterBitSet, cardinality, k));
+        }
```

Comment:
1. What if there are no engine files? I think you have to verify that doANNSearch results are not empty to make sure that ANN Search is success. 
2. Also can you please add Integration test?
3. if cardinality == maxDoc, why can't we pass null instead of filterBitSet?

Reply:
Already update the pull request and address the comments.